ostd/src/arch/riscv/irq/chip/plic.rs

```diff
@@ -0,0 +1,358 @@
+
+use alloc::{boxed::Box, collections::btree_map::BTreeMap, vec::Vec};
+
+use bit_field::BitField;
+use fdt::Fdt;
+
+use crate::{
+    io::{IoMem, IoMemAllocatorBuilder, Sensitive},
+    irq::IrqLine,
+    Error, Result,
+};
+
+/// The Platform-Level Interrupt Controller (PLIC) for RISC-V.
+pub(super) struct Plic {
+    phandle: u32,
+    io_mem: IoMem<Sensitive>,
+    hart_to_target_mapping: BTreeMap<u32, u32>,
+    /// Per-Plic interrupt-source-to-IRQ-number mappings.
+    interrupt_number_mappings: Box<[Option<u8>]>,
+}
+
+impl Plic {
+    pub(super) fn phandle(&self) -> u32 {
+        self.phandle
+    }
+
+    pub(super) fn interrupt_number_mapping(&self, interrupt_source: u32) -> Option<u8> {
+        self.interrupt_number_mappings[interrupt_source as usize]
+    }
+
+    fn num_interrupt_sources(&self) -> u32 {
+        self.interrupt_number_mappings.len() as u32
+    }
+
+    /// Sets the priority of an interrupt source.
+    pub(super) fn set_priority(&mut self, interrupt_source: u32, priority: u32) {
+        assert!(interrupt_source > 0 && interrupt_source < self.num_interrupt_sources());
+
+        let offset = Self::PRIORITY_OFFSET + 4 * interrupt_source as usize;
+        // SAFETY: The calculation of `offset` follows RISC-V PLIC's
+        // specification and is guaranteed to be valid.
+        // Reference: <https://github.com/riscv/riscv-plic-spec>.
+        unsafe { self.io_mem.write_once(offset, &priority).unwrap() };
+    }
+
+    /// Checks if an interrupt source is pending.
+    pub(super) fn is_pending(&self, interrupt_source: u32) -> bool {
+        assert!(interrupt_source > 0 && interrupt_source < self.num_interrupt_sources());
+
+        let word_index = interrupt_source as usize / 32;
+        let bit_index = interrupt_source as usize % 32;
+        let offset = Self::PENDING_OFFSET + 4 * word_index;
+        // SAFETY: The calculation of `offset` follows RISC-V PLIC's
+        // specification and is guaranteed to be valid.
+        // Reference: <https://github.com/riscv/riscv-plic-spec>.
+        let value = unsafe { self.io_mem.read_once::<u32>(offset).unwrap() };
+        (value >> bit_index) & 1 != 0
+    }
+
+    /// Sets whether an interrupt source is enabled for a specific target (hart).
+    pub(super) fn set_interrupt_enabled(&self, hart: u32, interrupt_source: u32, enabled: bool) {
+        assert!(interrupt_source > 0 && interrupt_source < self.num_interrupt_sources());
+
+        let target = *self.hart_to_target_mapping.get(&hart).unwrap();
+        let word_index = interrupt_source as usize / 32;
+        let bit_index = interrupt_source as usize % 32;
+        let offset = Self::ENABLE_OFFSET + Self::ENABLE_STRIDE * target as usize + 4 * word_index;
+
+        let mut value = unsafe { self.io_mem.read_once::<u32>(offset).unwrap() };
+        value.set_bit(bit_index, enabled);
+
+        // SAFETY: The calculation of `offset` follows RISC-V PLIC's
```

Comment:
Move the safety comment above and delete the blank line, since the comment applies to the two `unsafe` blocks:
```rust
        // SAFETY: The calculation of `offset` follows RISC-V PLIC's
        // specification and is guaranteed to be valid.
        // Reference: <https://github.com/riscv/riscv-plic-spec>.
        let mut value = unsafe { self.io_mem.read_once::<u32>(offset).unwrap() };
        value.set_bit(bit_index, enabled);
        unsafe { self.io_mem.write_once(offset, &value).unwrap() };
```